File: Data Structures & Algorithms/reorder-linked-list/submission-0.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

public class Solution {
    public void ReorderList(ListNode head) {
        var to_rev = head;
        ListNode rev = null;
        var fast = head;
        var slow = head;
        int counter = 1;
        while (fast != null) {
            if (counter == 0) {
                counter = 1;
                slow = slow.next;
            }
            else {
                counter -= 1;
            }
            fast = fast.next;
        }
        var swap = slow.next;
        slow.next = null;
        slow = swap;
        // Console.WriteLine(slow.val + " " + prev_slow.val);
        // reverse second 
        while (slow != null) {
            var curr = slow;
            slow = slow.next;
            curr.next = rev;
            rev = curr;
        }
        int step = 0;
        var iter = head;
        while (iter != null) {
            if ((step & 1) == 0) {
                if (rev == null) {
                    break;
                }
                var append = iter.next;
                iter.next = rev;
                rev = rev.next;
                iter.next.next = append;
            }
            iter = iter.next;
            step += 1;
        }
    }
}

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-0.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//     pub val: i32,
//     pub next: Option<Box<ListNode>>,
// }
//
// impl ListNode {
//     #[inline]
//     pub fn new(val: i32) -> Self {
//         ListNode { next: None, val }
//     }
// }

impl Solution {
    pub fn remove_nth_from_end(head: Option<Box<ListNode>>, n: i32) -> Option<Box<ListNode>> {
        let mut counter: i32 = 0;
        // separate scope so that we drop a reference
        {
            let mut fast: &Option<Box<ListNode>> = &head;
            while let Some(node_fast) = fast.as_ref() {
                fast = &node_fast.next;
                counter += 1;
            }
        }
        let mut dum = Some(Box::new(ListNode { val: 0, next: head }));
        let mut iter: &mut Option<Box<ListNode>> = &mut dum;
        // counter = length of the list
        // index of last element == length - 1
        // index of element to remove would be length - n
        let mut i = 0;
        let counter = counter;
        while i < counter - n {
            iter = &mut iter.as_mut().unwrap().next;
            i += 1;
        }
        if let Some(mut thing) = iter.as_mut() {
            let gap = thing.next.take();
            thing.next = gap.unwrap().next;
        }
        dum.unwrap().next
    }
}
